`background-repeat` is incorrectly exposed through inline style

smfr opened the pull request:
#### 165b098dbd0b11a2784a957b15265545b12dcfc8
<pre>
`background-repeat` is incorrectly exposed through inline style
<a href="https://bugs.webkit.org/show_bug.cgi?id=243915">https://bugs.webkit.org/show_bug.cgi?id=243915</a>

Reviewed by Antti Koivisto.

Remove the hack that treated `background-repeat` and `mask-repeat` as shorthands
for -x and -y longhands, since no such longhands are specified. This hack leaked
out via `element.style.length`: setting a `background-repeat` value would internally
set `background-repeat-x` and `background-repeat-y`, but then we&apos;d consider those
internal, and end up reporting `element.style.length` as zero. This broke the mechanism
used by <a href="https://css3test.com/">https://css3test.com/</a> to detect support for style properties.

`background-repeat` has custom serialization behavior: `repeat no-repeat` is serialized
as `repeat-x` for example. To support this, introduce a new CSSValue type CSSBackgroundRepeatValue.

CSSPropertyBackgroundRepeatX/Y and CSSPropertyMaskRepeatX/Y are gone, and various bits of
code treating CSSPropertyBackgroundRepeat and CSSPropertyMaskRepeat as shorthands is removed.

FillLayer stores a FillRepeatXY struct with x and y members.

* LayoutTests/fast/backgrounds/repeat/parsing-background-repeat-expected.txt:
* LayoutTests/fast/backgrounds/repeat/parsing-background-repeat.html: Add some tests for `style.length`. Adjust for new serialization behavior.
* LayoutTests/fast/css/background-layers-initial-size.html: Apply style to just the test div.
* LayoutTests/fast/css/remove-shorthand-expected.txt: Ordering changed.
* LayoutTests/fast/masking/parsing-mask-expected.txt:
* LayoutTests/fast/masking/parsing-mask-repeat.html: Added.
* LayoutTests/fast/masking/parsing-mask.html: Remove unnecessary font size style. `repeat no-repeat` serializes as `repeat-x` now.
* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt:
* LayoutTests/fast/masking/parsing-webkit-mask.html: Remove unnecessary font size style. `repeat no-repeat` serializes as `repeat-x` now.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSBackgroundRepeatValue.cpp: Added.
(WebCore::CSSBackgroundRepeatValue::CSSBackgroundRepeatValue):
(WebCore::CSSBackgroundRepeatValue::customCSSText const):
(WebCore::CSSBackgroundRepeatValue::equals const):
* Source/WebCore/css/CSSBackgroundRepeatValue.h: Added.
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::fillRepeatToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillRepeat):
(WebCore::CSSToStyleMap::mapFillRepeatX): Deleted.
(WebCore::CSSToStyleMap::mapFillRepeatY): Deleted.
* Source/WebCore/css/CSSToStyleMap.h:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::equals const):
(WebCore::CSSValue::cssText const):
(WebCore::CSSValue::destroy):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isBackgrouneRepeatValue const):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::StyleProperties::getLayeredShorthandValue const):
(WebCore::StyleProperties::asTextInternal const):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isValidKeywordPropertyAndValue):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeRepeatStyleComponent):
(WebCore::consumeRepeatStyle):
(WebCore::consumeBackgroundComponent):
(WebCore::CSSPropertyParser::parseSingleValue):
(WebCore::CSSPropertyParser::consumeBackgroundShorthand):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/display/css/DisplayFillLayerImageGeometry.cpp:
(WebCore::Display::geometryForLayer):
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::calculateExtendedBackgroundMode const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::calculateBackgroundImageGeometry const):
* Source/WebCore/rendering/style/FillLayer.cpp:
(WebCore::FillLayer::FillLayer):
(WebCore::FillLayer::operator=):
(WebCore::FillLayer::operator== const):
(WebCore::FillLayer::fillUnsetProperties):
(WebCore::FillLayer::hasRepeatXY const):
(WebCore::operator&lt;&lt;):
* Source/WebCore/rendering/style/FillLayer.h:
(WebCore::FillRepeatXY::operator== const):
(WebCore::FillLayer::repeat const):
(WebCore::FillLayer::isRepeatSet const):
(WebCore::FillLayer::setRepeat):
(WebCore::FillLayer::clearRepeat):
(WebCore::FillLayer::initialFillRepeat):
(WebCore::FillLayer::repeatX const): Deleted.
(WebCore::FillLayer::repeatY const): Deleted.
(WebCore::FillLayer::isRepeatXSet const): Deleted.
(WebCore::FillLayer::isRepeatYSet const): Deleted.
(WebCore::FillLayer::setRepeatX): Deleted.
(WebCore::FillLayer::setRepeatY): Deleted.
(WebCore::FillLayer::clearRepeatX): Deleted.
(WebCore::FillLayer::clearRepeatY): Deleted.
(WebCore::FillLayer::initialFillRepeatX): Deleted.
(WebCore::FillLayer::initialFillRepeatY): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::isIdempotentTextAutosizingCandidate const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::backgroundRepeat const):
(WebCore::RenderStyle::maskRepeat const):
(WebCore::RenderStyle::setBackgroundRepeat):
(WebCore::RenderStyle::setMaskRepeat):
(WebCore::RenderStyle::backgroundRepeatX const): Deleted.
(WebCore::RenderStyle::backgroundRepeatY const): Deleted.
(WebCore::RenderStyle::maskRepeatX const): Deleted.
(WebCore::RenderStyle::maskRepeatY const): Deleted.
(WebCore::RenderStyle::setBackgroundRepeatX): Deleted.
(WebCore::RenderStyle::setBackgroundRepeatY): Deleted.
(WebCore::RenderStyle::setMaskRepeatX): Deleted.
(WebCore::RenderStyle::setMaskRepeatY): Deleted.

Canonical link: <a href="https://commits.webkit.org/253440@main">https://commits.webkit.org/253440@main</a>
</pre>

nt1m commented on Source/WebCore/css/CSSValue.h:
```diff
@@ -89,6 +89,7 @@ class CSSValue {
         
 
     bool isAspectRatioValue() const { return m_classType == AspectRatioClass; }
+    bool isBackgrouneRepeatValue() const { return m_classType == BackgroundRepeatClass; }
```
```suggestion
    bool isBackgroundRepeatValue() const { return m_classType == BackgroundRepeatClass; }
```

typo

anttijk commented on Source/WebCore/css/CSSBackgroundRepeatValue.cpp:
```diff
@@ -0,0 +1,62 @@
+
+namespace WebCore {
+
+CSSBackgroundRepeatValue::CSSBackgroundRepeatValue(Ref<CSSPrimitiveValue>&& repeatXValue, Ref<CSSPrimitiveValue>&& repeatYValue)
```
Do we really need another trivial CSSValue for this? I think many cases like this are handled by a Pair type CSSPrimitiveValue ("offset-position", "border-top-left-radius" etc).

smfr replied:
It's hard to do custom serialization for a pair type everywhere it's needed, like in the `background` shorthand (I started down this route).

HerculesH committed:
[Gardening]: REGRESSION(253309@main?): [ iOS ] imported/w3c/web-platform-tests/html/semantics/forms/form-submission-0/implicit-submission.optional.html is a constant failure
https://bugs.webkit.org/show_bug.cgi?id=243946

Unreviewed test gardening.

* LayoutTests/platform/ios/TestExpectations:

Canonical link: https://commits.webkit.org/253433@main

smfr committed:
`background-repeat` is incorrectly exposed through inline style
https://bugs.webkit.org/show_bug.cgi?id=243915

Reviewed by Antti Koivisto.

Remove the hack that treated `background-repeat` and `mask-repeat` as shorthands
for -x and -y longhands, since no such longhands are specified. This hack leaked
out via `element.style.length`: setting a `background-repeat` value would internally
set `background-repeat-x` and `background-repeat-y`, but then we'd consider those
internal, and end up reporting `element.style.length` as zero. This broke the mechanism
used by https://css3test.com/ to detect support for style properties.

`background-repeat` has custom serialization behavior: `repeat no-repeat` is serialized
as `repeat-x` for example. To support this, introduce a new CSSValue type CSSBackgroundRepeatValue.

CSSPropertyBackgroundRepeatX/Y and CSSPropertyMaskRepeatX/Y are gone, and various bits of
code treating CSSPropertyBackgroundRepeat and CSSPropertyMaskRepeat as shorthands is removed.

FillLayer stores a FillRepeatXY struct with x and y members.

* LayoutTests/fast/backgrounds/repeat/parsing-background-repeat-expected.txt:
* LayoutTests/fast/backgrounds/repeat/parsing-background-repeat.html: Add some tests for `style.length`. Adjust for new serialization behavior.
* LayoutTests/fast/css/background-layers-initial-size.html: Apply style to just the test div.
* LayoutTests/fast/css/remove-shorthand-expected.txt: Ordering changed.
* LayoutTests/fast/masking/parsing-mask-expected.txt:
* LayoutTests/fast/masking/parsing-mask-repeat.html: Added.
* LayoutTests/fast/masking/parsing-mask.html: Remove unnecessary font size style. `repeat no-repeat` serializes as `repeat-x` now.
* LayoutTests/fast/masking/parsing-webkit-mask-expected.txt:
* LayoutTests/fast/masking/parsing-webkit-mask.html: Remove unnecessary font size style. `repeat no-repeat` serializes as `repeat-x` now.
* Source/WebCore/Sources.txt:
* Source/WebCore/WebCore.xcodeproj/project.pbxproj:
* Source/WebCore/animation/CSSPropertyAnimation.cpp:
(WebCore::CSSPropertyAnimationWrapperMap::CSSPropertyAnimationWrapperMap):
* Source/WebCore/css/CSSBackgroundRepeatValue.cpp: Added.
(WebCore::CSSBackgroundRepeatValue::CSSBackgroundRepeatValue):
(WebCore::CSSBackgroundRepeatValue::customCSSText const):
(WebCore::CSSBackgroundRepeatValue::equals const):
* Source/WebCore/css/CSSBackgroundRepeatValue.h: Added.
* Source/WebCore/css/CSSComputedStyleDeclaration.cpp:
(WebCore::fillRepeatToCSSValue):
(WebCore::ComputedStyleExtractor::valueForPropertyInStyle):
* Source/WebCore/css/CSSProperties.json:
* Source/WebCore/css/CSSToStyleMap.cpp:
(WebCore::CSSToStyleMap::mapFillRepeat):
(WebCore::CSSToStyleMap::mapFillRepeatX): Deleted.
(WebCore::CSSToStyleMap::mapFillRepeatY): Deleted.
* Source/WebCore/css/CSSToStyleMap.h:
* Source/WebCore/css/CSSValue.cpp:
(WebCore::CSSValue::equals const):
(WebCore::CSSValue::cssText const):
(WebCore::CSSValue::destroy):
* Source/WebCore/css/CSSValue.h:
(WebCore::CSSValue::isBackgrouneRepeatValue const):
* Source/WebCore/css/StyleProperties.cpp:
(WebCore::StyleProperties::getPropertyValue const):
(WebCore::StyleProperties::getLayeredShorthandValue const):
(WebCore::StyleProperties::asTextInternal const):
* Source/WebCore/css/parser/CSSParserFastPaths.cpp:
(WebCore::CSSParserFastPaths::isValidKeywordPropertyAndValue):
* Source/WebCore/css/parser/CSSPropertyParser.cpp:
(WebCore::consumeRepeatStyleComponent):
(WebCore::consumeRepeatStyle):
(WebCore::consumeBackgroundComponent):
(WebCore::CSSPropertyParser::parseSingleValue):
(WebCore::CSSPropertyParser::consumeBackgroundShorthand):
(WebCore::CSSPropertyParser::parseShorthand):
* Source/WebCore/display/css/DisplayFillLayerImageGeometry.cpp:
(WebCore::Display::geometryForLayer):
* Source/WebCore/page/FrameView.cpp:
(WebCore::FrameView::calculateExtendedBackgroundMode const):
* Source/WebCore/rendering/RenderBoxModelObject.cpp:
(WebCore::RenderBoxModelObject::calculateBackgroundImageGeometry const):
* Source/WebCore/rendering/style/FillLayer.cpp:
(WebCore::FillLayer::FillLayer):
(WebCore::FillLayer::operator=):
(WebCore::FillLayer::operator== const):
(WebCore::FillLayer::fillUnsetProperties):
(WebCore::FillLayer::hasRepeatXY const):
(WebCore::operator<<):
* Source/WebCore/rendering/style/FillLayer.h:
(WebCore::FillRepeatXY::operator== const):
(WebCore::FillLayer::repeat const):
(WebCore::FillLayer::isRepeatSet const):
(WebCore::FillLayer::setRepeat):
(WebCore::FillLayer::clearRepeat):
(WebCore::FillLayer::initialFillRepeat):
(WebCore::FillLayer::repeatX const): Deleted.
(WebCore::FillLayer::repeatY const): Deleted.
(WebCore::FillLayer::isRepeatXSet const): Deleted.
(WebCore::FillLayer::isRepeatYSet const): Deleted.
(WebCore::FillLayer::setRepeatX): Deleted.
(WebCore::FillLayer::setRepeatY): Deleted.
(WebCore::FillLayer::clearRepeatX): Deleted.
(WebCore::FillLayer::clearRepeatY): Deleted.
(WebCore::FillLayer::initialFillRepeatX): Deleted.
(WebCore::FillLayer::initialFillRepeatY): Deleted.
* Source/WebCore/rendering/style/RenderStyle.cpp:
(WebCore::RenderStyle::isIdempotentTextAutosizingCandidate const):
* Source/WebCore/rendering/style/RenderStyle.h:
(WebCore::RenderStyle::backgroundRepeat const):
(WebCore::RenderStyle::maskRepeat const):
(WebCore::RenderStyle::setBackgroundRepeat):
(WebCore::RenderStyle::setMaskRepeat):
(WebCore::RenderStyle::backgroundRepeatX const): Deleted.
(WebCore::RenderStyle::backgroundRepeatY const): Deleted.
(WebCore::RenderStyle::maskRepeatX const): Deleted.
(WebCore::RenderStyle::maskRepeatY const): Deleted.
(WebCore::RenderStyle::setBackgroundRepeatX): Deleted.
(WebCore::RenderStyle::setBackgroundRepeatY): Deleted.
(WebCore::RenderStyle::setMaskRepeatX): Deleted.
(WebCore::RenderStyle::setMaskRepeatY): Deleted.

Canonical link: https://commits.webkit.org/253440@main
